cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/CruiseControlMetricsProcessor.java

```diff
@@ -292,449 +262,355 @@ private TopicPartition partitionHandleDotInTopicName(TopicPartition tp) {
-      return _topicIOLoad.compute(topicWithDotHandled, (t, l) -> {
-        if (l != null) {
-          return l;
+    /**
+     * Due to the yammer metric exponential decaying mechanism, the broker metric and the sum of the partition metrics
+     * on the same broker may differ by a lot. Our experience shows that in that case, the sum of the topic/partition
+     * level metrics are more accurate. So we will just replace the following metrics with the sum of topic/partition
+     * level metrics:
+     * <ul>
+     *   <li>BrokerProduceRate</li>
+     *   <li>BrokerFetchRate</li>
+     *   <li>BrokerLeaderBytesInRate</li>
+     *   <li>BrokerLeaderBytesOutRate</li>
+     *   <li>BrokerReplicationBytesInRate (if exists)</li>
+     *   <li>BrokerReplicationBytesOutRate (if exists)</li>
+     *   <li>BrokerMessagesInRate</li>
+     * </ul>
+     *
+     * We use the cluster metadata to check if the reported topic level metrics are complete. If the reported topic
+     * level metrics are not complete, we ignore the broker metric sample by setting the valid flat to false.
+     *
+     * @param cluster The Kafka cluster.
+     * @param brokerId the broker id to prepare metrics for.
+     * @param time the last sample time.
+     */
+    private void prepareBrokerMetrics(Cluster cluster, int brokerId, long time) {
+      boolean enoughTopicPartitionMetrics = enoughTopicPartitionMetrics(cluster, brokerId);
+      // Ensure there are enough topic level metrics.
+      if (enoughTopicPartitionMetrics) {
+        Map<RawMetricType, Double> sumOfTopicMetrics = new HashMap<>();
+        for (String topic : _topicsWithPartitionSizeReported) {
+          METRIC_TYPES_TO_SUM.keySet().forEach(type -> {
+            double value = topicMetrics(topic, type, false);
+            sumOfTopicMetrics.compute(type, (t, v) -> (v == null ? 0 : v) + value);
+          });
+        }
+        for (Map.Entry<RawMetricType, Double> entry : sumOfTopicMetrics.entrySet()) {
+          RawMetricType rawTopicMetricType = entry.getKey();
+          double value = entry.getValue();
+          _brokerMetrics.setRawMetricValue(METRIC_TYPES_TO_SUM.get(rawTopicMetricType), value, time);
+        }
+      }
+      // Check if all the broker raw metrics are available.
+      for (RawMetricType rawBrokerMetricType : RawMetricType.brokerMetricTypes()) {
+        if (_brokerMetrics.metricValue(rawBrokerMetricType) == null) {
+          _missingBrokerMetrics.add(rawBrokerMetricType);
+        }
+      }
+      // A broker metric is only available if it has enough valid topic metrics and it has reported
+      // replication bytes in/out metrics.
+      _brokerMetricsAvailable = enoughTopicPartitionMetrics && _missingBrokerMetrics.isEmpty();
+    }
+
+    /**
+     * Verify whether we have collected enough metrics to generate the broker metric samples. The broker must have
+     * collected more than 99% of the topic level and partition level metrics in the broker to generate broker
```

Comment:
Can we make this either 1) configurable or 2) more explicit  --e.g. via defining it as a constant?